### ISSUE TITLE
Snakemake rule does not regenerate output after input has changed

### ISSUE THREAD

**Issue by dariober** [snakemake/snakemake#2011]
**Snakemake version**: tested on 7.18.2 and 7.19.1

**Describe the bug**

Snakemake doesn't re-run the pipeline even if input file is newer than output. Version 6.15.5 behaves as expected (I haven't tested other versions). This is a simplified version of this question on [SO](https://stackoverflow.com/questions/74795939/snakemake-rule-does-not-regenerate-output-after-input-has-changed).

_Unless I'm missing something this looks like an urgent bug to fix?_

**Minimal example**

Minimal snakefile:

```
rule all:
    input:
        'out.txt'

rule compose:
    input:
        'in.txt',
    output:
        'out.txt',
    shell:
        r"""
        touch {output}
        """
```

**Steps to reproduce**

* Create input file and run pipeline (so far so good):

```
touch in.txt
snakemake -p -j 1

snakemake -p -j 1
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job        count    min threads    max threads
-------  -------  -------------  -------------
all            1              1              1
compose        1              1              1
total          2              1              1
...
```

* Update input, rerun pipeline, and you get `nothing to be done`:

```
rm in.txt 
touch in.txt
snakemake -p -j 1
Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).
Complete log: .snakemake/log/2022-12-14T105436.335608.snakemake.log

# Chekc input newer than output:
stat in.txt out.txt
  File: in.txt
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 802h/2050d	Inode: 3153450     Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1001/   dario)   Gid: ( 1001/   dario)
Access: 2022-12-14 11:05:41.726815101 +0000
Modify: 2022-12-14 11:05:41.726815101 +0000
Change: 2022-12-14 11:05:41.726815101 +0000
 Birth: -
  File: out.txt
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 802h/2050d	Inode: 3177457     Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1001/   dario)   Gid: ( 1001/   dario)
Access: 2022-12-14 11:04:28.938495477 +0000
Modify: 2022-12-14 11:04:28.938495477 +0000
Change: 2022-12-14 11:04:28.938495477 +0000
 Birth: -

```

* Use snakemake 6.15.5 and you do get a new execution:

```
snakemake -v
6.15.5

snakemake -p -j 1 
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job stats:
job        count    min threads    max threads
-------  -------  -------------  -------------
all            1              1              1
compose        1              1              1
total          2              1              1
...
```

**Comment by tbooth** [comment]
Yep, exact same behaviour for me. It's not your imagination @dariober - Snakemake is very broken until this gets sorted out! I'm not sure how it's even passing the basic test suite right now?

**Comment by tbooth** [comment]
I did a bit more poking. If the actual content of the file is changed then Snakemake re-runs as expected. If the .snakemake directory is deleted then Snakemake also re-runs as expected.

The root of the issue looks to be in `snakemake/dag.py` line 1124:

~~~
        # Input is updated if it is newer that the oldest output file
        # and does not have the same checksum as the one previously recorded.
        updated_input = [
             f
             for f in job.input
             if f.exists
             and f.is_newer(output_mintime_)
             and not is_same_checksum(f, job)
        ]
~~~

My first thought was that the final `and` should be an `or` because we want to re-run the rule if the input is newer **or** if the contents of the file seem have changed, but the comment makes it clear that the behaviour is as-intended.

I guess this makes sense. Snakemake generally re-runs if the input file is newer, but makes an exception if it has a checksum in its database showing that the file contents is actually unchanged. Older Snakemake is very rerun-happy so this could save a good amount of unnecessary re-running.

Likewise it will not trigger the rule if the time is newer even if the content seems to have changed. This is necessary for the `touch` logic to operate as advertised.

However...

For one thing there's no documentation I can see that explains this and it goes against most people's assumptions of how Snakemake works. It also breaks any logic using touch files as surrogate inputs for example.

The setting for `--rerun-triggers=mtime` does not modify this logic, even though the docs say it will restore the "traditional" behaviour.  I would expect the `--rerun-triggers=mtime` setting to disable any attempt to checksum file contents at all. On a tiered storage system, reading a load of input files to generate checksums could cause a massive slowdown, even though Snakemake imposes an arbitrary limit of 100kb (see io.py line 587). Even on a regular local FS it's a significant waste of time if not actually needed. So I think the real fix involves bypassing file checksum calculation phase based on the `--rerun-triggers=mtime` setting, and then `is_same_checksum(f, job)` will always return `False` and so the logic would be correct and depend only on `f.is_newer()`.

Final comment - the code for `is_same_checksum()` looks pretty weird. I really don't understand what this is up to, but I guess it seems to work as advertised!

**Comment by liao961120** [comment]
I tested various versions and found that version **7.9.0** worked as expected (touching the dependencies results in re-running).
Version **7.10.0** and above behaved as what @tbooth had described:
> [...] If the actual content of the file is changed then Snakemake re-runs as expected. If the .snakemake directory is deleted then Snakemake also re-runs as expected.

**Comment by mapio** [comment]
This is really bad and confusing. Today I spent hours trying to debug a complex flow that was not generating missing file.

The problem turned out that a rerun isn't triggered not even if a file is missing!

Consider this flow

```
rule second:
  input: 'two.txt'
  output: 'three.txt'
  shell: 'cat {input} > {output}'
```

if you `touch one.txt` and run it, the file `three.txt` is correctly generated. 

Now if you touch again `one.txt` or even *delete* it, nothing happens when you launch `snakemake`.

In this simple example, I've noticed that `rm -f .snakemake/metadata*` seems to solve the problem (if you touch or delete the file and then remove the metadata, the flow reruns as expected, if you remove the metadata and leave the file unchanged, the flow does not run).

But I'm honestly scared of deleting the metadata of the large and complex flow I'm debugging because I don't want to trigger expensive recomputations.

Is there a plan to fix this?

**Comment by mapio** [comment]
Seems to be related to https://github.com/snakemake/snakemake/issues/1978 where the thing appears to be considered a feature instead of a bug.

**Comment by mapio** [comment]
And bisecting versions, I've found that this happens in 7.7 but not in 7.6.2.
